VideoCommon: add a graphics mod action that allows you to modify the game's base rendering

iwubcode opened the pull request:
Graphics mods are pretty basic right now.  Disable bloom or stretch a HUD element.  Not very exciting!  But graphics mods were meant to be so much more.

With this feature, Dolphin gives users _direct_ modification of the way game's draw by exposing hook points for users to provide their own pixel shaders.

### A Basic Example

If you wanted to turn a character red, you'd provide a graphics mods that targeted that character's texture.  For example, take Prince of Persia: the Forgotten Sand's main character texture here's how we might target that with our new graphics mod:

```
{
	"meta":
	{
		"title": "Color test pipeline",
		"author": "iwubcode"
	},
	"assets": [
		{
			"name": "material_color",
			"data":
			{
				"": "material.json"
			}
		},
		{
			"name": "shader_color",
			"data":
			{
				"metadata": "color.shader.json",
				"shader": "color.glsl"
			}
		}
	],
	"features": [
		{
			"action": "custom_pipeline",
			"action_data": {
				"passes": [
					{
						"pixel_material_asset": "material_color"
					}
				]
			},
			"group": "PrinceColor"
		}
	],
	"groups": [
		{
			"name": "PrinceColor",
			"targets": [
				{
					"texture_filename": "tex1_512x512_m_d0d80f2dcf9ed354_14",
					"type": "draw_started"
				},
				{
					"texture_filename": "tex1_512x512_m_d0d80f2dcf9ed354_14",
					"type": "create_texture"
				}
			]
		}
	]
}
```

with the color shader metadata ( `color.shader.json` ) looking like:

```
{
	"properties": []
}
```

the material json looking like ( `material.json` ):

```
{
	"shader_asset": "shader_color",
	"values": []
}
```

And finally the actual shader "color.glsl" that the user writes:

```
vec4 custom_main( in CustomShaderData data )
{
	return vec4(1.0, 0.0, 0.0, 1.0);
}
```

What's going on here?  Well Dolphin is taking this piece of code and returning the color red.  Astute readers might notice the `CustomShaderData`, more on that later.

If you're curious, here's the result:

![pop_tfs](https://user-images.githubusercontent.com/15224722/203684916-5a4c0975-d8bb-495f-aec8-984d1165869d.png)

### Game lighting

While it's up to the shader creator to decide what to do, the most obvious use case is to modify the game's lighting.  Let's talk about that!

Game's provide two things for lighting:  normals (a direction to describe the slope of the surface) and lights.  In Dolphin, when Per-Pixel-LIghting is turned on, Dolphin will provide the normal to its pixel shader.  This normal is then provided to the user through the `CustomShaderData`.  So if you wanted to display a game's normals, you'd simply do:

```
vec4 custom_main( in CustomShaderData data )
{
	return vec4(data.normal, 1);
}
```

Applied to a handful of Fragile Dreams: Farewell Ruins to the Moon's textures, you'd see:

![fragile-normals](https://user-images.githubusercontent.com/15224722/203678660-657cd1e9-455b-42ad-bc32-595ae25dfad7.png)


Normals are wonderful but we really want to do something with lighting.  In order to do that we need _lights_.  Dolphin provides you all the scene's lights in the `CustomShaderData`.  This includes their type, the color, and their position.  You also get the pixels position, meaning with a little math you have all you need to get some lighting outputted.

Viewing lighting by itself might look like this (Little King's Story):

![little king's story light](https://user-images.githubusercontent.com/15224722/203679257-9e604d01-3190-4631-9786-fb21d346205f.gif)

### Game Quirks

Alas.  Games don't have to give you normals.  Arc Rise Fantasia fails to provide any:

![arc rise fantasia - no normals](https://user-images.githubusercontent.com/15224722/203679403-89876329-d777-4244-8684-eee547f50343.png)

Some games give you normals but the view direction can modify them.  The Last Story is an example of that:

![last-story-changing_normals](https://user-images.githubusercontent.com/15224722/203680644-3987dbbf-375b-4025-9891-991e83235b4e.gif)

Some games give you normals but don't actually provide lights on those objects.  An example is Rune Factory: Tides of Destiny which provides normals on all objects but only provides lighting on the characters.  In those situations you have the option to bake in the lighting directly in the shader:

![rf_tod_shader_lighting](https://user-images.githubusercontent.com/15224722/203682408-5ab367ce-4d38-4158-a298-cf96bc9deff8.png)

I've only tested 20 - 30 games at the moment but a majority of them have lighting/normals!  Users will have to test their games to see what they can do with them.

### Leveraging Textures

We have the game's normals and we may have lights.  But the game's data is rather limited.  What we really want is to provide textures to our shaders to give it more information.  Well luckily we can!!!

Adding a texture to a shader requires us to use _assets_.  You might have noticed this in the Prince example above but a material, a shader, and that's right even textures are all assets!

To leverage a texture, we need to:

* Create a texture asset that references our texture file
* Create a shader asset that says we expose a texture sampler in our shader code
* Create a material that references our texture asset and syncs that with the texture sampler in the shader asset.

Whew!

Just showing the `assets` and `features` section it might look like this:

```
    "assets": [
        {
            "name": "material_output_normal",
            "data":
            {
                "": "output_normal.material.json"
            }
        },
        {
            "name": "output_replace_normal",
            "data":
            {
                "metadata": "output_normal.shader.json",
                "shader": "output_normal.glsl"
            }
        },
        {
            "name": "tex1_1024x512_m_8cfb7d28ee78826b_14_normal",
            "data":
            {
                "": "tex1_1024x512_m_8cfb7d28ee78826b_14_normal.png"
            }
        }
    ],
    "features": [
        {
            "action": "custom_pipeline",
            "action_data": {
              "passes":  [
                {
                  "pixel_material_asset": "material_output_normal"
                }
              ]
            },
            "group": "NormalShaderTarget_tex1_1024x512_m_8cfb7d28ee78826b_14"
        }
    ]
```

where `output_normal.material.json` would look like:

```
{
    "shader_asset": "shader_replace_normal",
    "values": [
        {
            "type": "texture_asset",
            "code_name": "NORMAL_TEX",
            "value": "tex1_1024x512_m_8cfb7d28ee78826b_14_normal"
        }
    ]
}
```

and `output_normal.shader.json` would look like:

```
{
    "properties": [
        {
            "type": "sampler2d",
            "description": "The normal texture",
            "code_name": "NORMAL_TEX"
        }
    ]
}
```

Finally, with all that in place, the `output_normal.glsl` might look like::

```
vec4 custom_main( in CustomShaderData data )
{
	if (data.texcoord_count == 0)
	{
		vec3 normal = data.normal.xyz;
		return vec4((normal * 0.5f ) + 0.5f, 1.0);
	}
	else
	{
#ifdef NORMAL_TEX_UNIT
		vec4 map_rgb = texture(samp[NORMAL_TEX_UNIT], NORMAL_TEX_COORD);
		vec4 map = map_rgb * 2.0 - 1.0;
		return map;
#else
		vec3 normal = data.normal.xyz;
		return vec4((normal * 0.5f ) + 0.5f, 1.0);
#endif
	}
}
```

This shader will either display the draw with the normal data (instead of the game's computed color) or will use the normal texture if available.

Where do `NORMAL_TEX_UNIT` and `NORMAL_TEX_COORD` come from?  Well, in the shader metadata you specified an input sampler of `NORMAL_TEX` and Dolphin automatically generates `NORMAL_TEX_UNIT` and `NORMAL_TEX_COORD` for you.

You aren't restricted to one texture however.  You can provide as many as you want following the same process.  This allows you to do interesting features like physically based rendering (PBR).  Here's Mobile Suit Gundam MS Sensen 0079 with a Mettalic/Roughness/Normal map:

![gundam-pbr](https://user-images.githubusercontent.com/15224722/203680874-e5f0291f-1eff-4947-acce-65b74c340c36.gif)


Parallax occlusion is also possible.

Default

![tos_parallax_occulsion_no](https://user-images.githubusercontent.com/15224722/209851777-a0a00658-4e79-41c8-8543-c8f8a1f4279b.png)

With parallax occlusion on (and metal details):

![tos_parallax_occulsion](https://user-images.githubusercontent.com/15224722/209851875-11d7829b-80b5-42a6-a771-d69dcc8fafdb.png)


A time value is exposed, so basic animation is possible:

![raguna_blink](https://user-images.githubusercontent.com/15224722/211138099-be3b8266-566d-473e-8847-c48487e91305.gif)



This all might be quite complicated.  And it is definitely very technical at the moment.  Creating the boiler plate data for assets, materials, and shaders is tedious.  I've written [some scripts](https://github.com/iwubcode/dolphin-graphic-mods-scripts) to make this simpler but all this complexity is expected to be alleviated with some sort of editor.  Even not taking the assets into consideration, it will take time for the community to get up to speed to write shaders that can be leveraged.

### Future work

In the future we can support even more features:

* We can force normals to be generated with either geometry or tessellation shaders.  This will allow games without normals (or broken normals) to have full use of the functionality
* We can allow users to add additional (and dynamic!) lights.  This requires knowing the position of elements, which means we need to look at hooking into the VertexLoader
* Cube map support for static reflections
* Expose shader compilation threading in a setting.
* Shader uniforms
* An editor to make it easier to create the various assets


---

Go out and make some pretty games!

Turn this:
![harvest-moon-no-shader](https://user-images.githubusercontent.com/15224722/203681655-5c4b79a4-f4b7-41da-aed9-548074a6e332.png)
Into this:
![harvest-moon-shader](https://user-images.githubusercontent.com/15224722/203681710-5e7209c8-2ea6-40a2-8ea8-94dc5a4e64b2.png)



![GRSEAF_2022-12-28_12-34-13](https://user-images.githubusercontent.com/15224722/209857518-57272809-f693-4443-95fa-edf826152be9.png)



![ffcccb](https://user-images.githubusercontent.com/15224722/203681767-59099b90-6977-4fe2-aa77-4ddb0fb116c9.gif)

tellowkrinkle commented:
> Finalize the design. I originally opted to allow the texture inputs to be added independently of the game's textures (or custom textures). This allows the textures to be any size but has the downside of competing with the game's texture samplers (we only have 8). We could be smarter about samplers but that's another task.. Now I use texture arrays but that forces users to use the same texture size. I need to decide if I want to expose both..

Where does limitation of 8 textures come from?  If it's from the render backends, that could be adjusted.

iwubcode commented:
> Where does limitation of 8 textures come from? If it's from the render backends, that could be adjusted.

I think 8 was chosen because that matches the GC/Wii hardware.  But yes, I looked at increasing it.  12 worked for my machine but increasing to 16 caused my GPU to spaz out.  From what I understand, graphics cards support a very small amount of samplers but many textures.  Hence my comment about improving our sampler handling.  Of course, for my usage, the texture array approach works fine.

JoshuaMKW commented:
Watching with great interest

iwubcode commented:
Ubershaders is implemented.  Vulkan black textures are fixed.  The code/commits are in a pretty good state.  I see one more thing to do there.

Sadly, Vulkan still being Vulkan.  Validation errors when trying to copy the base texture + load the additional data.  No visible impact as far as I can tell but it should probably still be resolved.

Finally, I have been debugging some flickering I see in Final Fantasy Crystal Chronicles: Crystal Bearers.  Turns out, every other frame the game draws the floor with two textures instead of one.  In my testing, I used both textures which means the second frame draw actually draws the second texture (when the floor is drawn with the first).  While I can simply ignore the second texture in this case, I think it may be beneficial to have some sort of tev stage api.  I was planning on doing that in a follow-up PR but I may see if I can implement it here.

iwubcode commented:
I've added a minimal tev stage api for both the specialized and uber shaders.  It allows you to get the input and output of each stage, along with analyzing the input type.

For instance, say you wanted to remove character shading from Tales of Symphonia in order to add your own shading.  First you'd isolate just the character texture.

So change this:

![tos_default](https://user-images.githubusercontent.com/15224722/209006552-35e37563-43c7-4f7d-98cc-270720251416.png)

to this:

![tos_no_shading](https://user-images.githubusercontent.com/15224722/209006608-08996552-88f0-4e06-bd4c-8a19ca293fc0.png)


.  A shader would look like this:

```
vec4 custom_main( in CustomShaderData data )
{
	if (data.tev_stage_count == 0)
	{
		return data.final_color;
	}
	else
	{
		vec4 final_color = data.final_color;
		uint texture_set = 0;
		for (uint i = 0; i < data.tev_stage_count; i++)
		{
                        // There are 4 color inputs
			for (uint j = 0; j < 4; j++)
			{
				if (data.tev_stages[i].input_color[j].input_type == CUSTOM_SHADER_TEV_STAGE_INPUT_TYPE_TEX && texture_set == 0)
				{
					final_color = float4(data.tev_stages[i].input_color[j].value, 1.0);
					texture_set = 1;
				}
			}
		}

		return final_color;
	}
}
````

Which would then allow you to apply your own shading.  Here's Tales of Symphonia with a more realistic lighting (+ rim light):

![tos_no_shading_custom_lighting](https://user-images.githubusercontent.com/15224722/209007329-55c26f94-52e6-4d39-82a1-8a2cd83cea2a.png)

iwubcode commented:
The coveted parallax occlusion is also possible.  Thanks to @phire for working through this with me:

Default

![tos_parallax_occulsion_no](https://user-images.githubusercontent.com/15224722/209851777-a0a00658-4e79-41c8-8543-c8f8a1f4279b.png)

With parallax occlusion on (and metal details):

![tos_parallax_occulsion](https://user-images.githubusercontent.com/15224722/209851875-11d7829b-80b5-42a6-a771-d69dcc8fafdb.png)

iwubcode commented:
Calling on the graphics masters. 

@K0bin / @TellowKrinkle / @Pokechu22  - would you all be willing to start looking this over?  It's mostly ready.

Some things to call out:

1)  All the user shader code is put into the same source shader.  This can cause conflicts if users use a function with the same signature and name.  In order to avoid that, I wrote a mini glsl parser that will find all the function names and prefix/suffix them with a unique identifier.  It's not ideal but it works.  I discussed with @phire about alternatives and there wasn't really anything (at least not usable on all our backends).
2)  Testing shaders is a bit painful at the moment.  First of all, every change you do needs a recompile (there aren't any uniforms).  Also, that Dolphin pop up gets frustrating fast.  I run into a similar issue in my post processing overhaul.  Will eventually need to get that into a proper UI...
3)  There is some duplicated code.
   (a)  The shader cache class just felt very unwieldy to me, so there is a bit duplicated in the `CustomShaderCache`.  I do think it's pretty clean however.
   (b)  There is texture load code (png/dds) that was pulled out of hires textures.  My eventual plan is to make custom textures a graphics mod, but that's not available yet.  I hope to tackle this cleanup in a future PR.
4)  There isn't a disk shader cache at the moment.  This was intentional initially because I thought it would be weird with how the user can toggle on/off these shaders.  But maybe these can be tagged and when a mod is disabled the cache is deleted?  Not something that needs to be figured out at the moment but wanted to mention it.
5) I tried to take a light touch to the shader logic and keep most of the custom code in one place.  Obviously the TEV API sort of made that difficult.  Also, it'd be nice if we could move more of the shader code into separate functions for better reuse.  I also considered mimicking the API for the actual Dolphin code, might make things more readable.  Not sure if it'd be slower though...

There is still an open issue with Vulkan where there is a validation error.  The error is:

>37:19:042 VideoBackends\Vulkan\VulkanContext.cpp:745 E[Host GPU]: Vulkan debug report: (Validation) Validation Error: [ UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout ] Object 0: handle = 0x19,b20,96f,820, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x4d,ae5,635 | vkQueueSubmit(): pSubmits[0].pCommandBuffers[0] command buffer VkCommandBuffer 0x19,b20,96f,820[] expects VkImage 0x2,b58,e50,000,007,49a[] (subresource: aspectMask 0x1 array layer 3, mip level 0) to be in layout VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL--instead, current layout is VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL.

I do not know how to resolve it.  It happens during the point [where the game texture data is copied into a new texture so that the additional textures specified by the shader can be added as additional layers](https://github.com/iwubcode/dolphin/blob/custom-shaders/Source/Core/VideoCommon/TextureCacheBase.cpp#L1276).

Final thing I'll say is that the API was meant to be all encompassing so users wouldn't use anything in the shader header that we could change.  _However_ it doesn't include the variable _samp_ .  Apparently copying _samp_ causes some issues in SPIRV-Cross.  I can't recall the error offhand.

TellowKrinkle commented on Source/Core/VideoCommon/GraphicsModSystem/Runtime/CustomTextureData.cpp:
```diff
@@ -0,0 +1,539 @@
+
+namespace
+{
+// From https://raw.githubusercontent.com/Microsoft/DirectXTex/master/DirectXTex/DDS.h
```
Could this be combined with the DDS loader in `HiresTextures_DDSLoader.cpp`?

TellowKrinkle replied:
Oh never mind you mentioned that in 3b

TellowKrinkle reviewed (commented):
It would be nice if you could post an example of a custom shader and the GLSL (uber and non-uber) it gets converted to by the graphics mod, for reviewers who are having trouble reading the shader generation code.

iwubcode commented:
@TellowKrinkle - great idea.

For a user shader outlined in the opening (the prince of persia red shader), I did what you requested.  I introduced an error to get it to dump (removed a semicolon) but you should be able to follow it, let me know if you have any questions.

Here is the user shader:

```
vec4 custom_main( in CustomShaderData data )
{
  return vec4(1.0, 0.0, 0.0, 1.0)
}
```

Here is the output:

[specialized_shader.txt](https://github.com/dolphin-emu/dolphin/files/10328494/specialized_shader.txt)
[uber_shader.txt](https://github.com/dolphin-emu/dolphin/files/10328510/uber_shader.txt)

If you want it, I can get a lighting/texture example as well.

iwubcode commented:
I forgot one of the original reasons I wanted to implement this was to support animation.  Well that's now possible:

![raguna_blink](https://user-images.githubusercontent.com/15224722/211138099-be3b8266-566d-473e-8847-c48487e91305.gif)


Otherwise I cleaned up the code a bit more in the CustomShaderAction.

iwubcode commented:
Minor cleanup and fixed a possibility of a crash (null pixel shader in the pipeline when the shader fails to compile).

K0bin commented:
How do I reproduce the Vulkan validation error?

iwubcode commented:
@K0bin - thank you very much, you fixed the last blocker to this PR!  I didn't think of using the _existing_ layout as a check, very smart!

> Also, is it at all possible to maybe split this PR into smaller ones? It's practically impossible to review.

The first six or seven commits could be potentially split but most of the actual custom shader ones, need to be together.

iwubcode commented:
Is there something wrong with fifoci?  I would not expect my code to have any impact on rendering (in particular only on mac..)

iwubcode commented:
Yeah, on second thought, I think the fifoci error is caused by the change @K0bin suggested.  Must be some scenario where `Load` is being called on a non-new texture when it was expected to use the `InitCommandBuffer` and now I've changed it to use the `CurrentCommandBuffer`.

I'll either have to track down those scenarios and see if they are fixable or opt for another approach...

iwubcode commented:
I removed the validation errors from the todo list.  The validation errors don't seem to work anymore (or the error is gone?), the only indication is:

```
Vulkan debug message: Failed to find vkGetDeviceProcAddr in layer
```

it used to work (a few weeks ago), not sure what happened.

I removed the offending commit that was causing the macos fifoci errors.

K0bin commented:
That error seems like a setup issue with the Vulkan validation layers.

As to why my patch broke things: we probably need to end the render pass before doing the copy and resume it afterwards.

iwubcode commented:
Swapped around the ordering of how the logic works, now I load the new custom textures _first_ and then copy the old texture followed by a `FinishedRendering`.  This seems to resolve the validation error without impacting the base renderer logic which was causing the issue and could be tricky to fix in a performant way.

K0bin commented:
Did you make sure your Vulkan SDK is up to date? Also, does the validation error say which layer is the problem?
Apparently it's caused by broken Vulkan layers and usually tells you which one. https://stackoverflow.com/questions/63739366/vulkan-validation-layer-loader-create-instance-chain-failed-to-find-vkgetinst

iwubcode commented:
Vulkan SDK is 1.3.231

And yes, it gave a path to an AMD specific file.

K0bin commented:
*sigh* Maybe we need to #ifdef this for Android or disable it on AMD systems.

iwubcode commented:
Maybe.  Like I mentioned, I will try a reinstall first.  If we need to, let's continue any further conversation on #11219 or on Discord.

iwubcode commented:
No major changes, just renamed `CustomShaderAction` to `CustomPixelShaderAction` and removed a tiny function that was  unused.  Still ready for review.

JoshuaMKW commented:
What's currently left before this would get merged? It seems pretty complete if not just in the cleanup phase.

AdmiralCurtiss commented:
How much does this confict with @phire's continued refactoring? I'm thinking that if you both continue working on your respective things I'd rather get them merged sooner rather than later, just to avoid annoying rebases later.

iwubcode commented:
@JoshuaMKW  - main thing is getting a review.

I have been tinkering with an idea of wildcards for graphics mods.  That has been very helpful for debugging my entire library (previously I had to manually make a mod for each game).  I just pushed up a few bug fixes after testing 50-70 games.  There may be more bugs or even an API change needed but at it is functional.

As I mentioned on Discord, it may be _functional_ but it is kinda a pain from a usability perspective.  After this is merged, I will focus on a few other areas before tackling usability.  There's still a lot to do.  This is step number 1.

@AdmiralCurtiss - thank you.  I think most of the conflicting parts have been broken off and merged.  There are changes to the pixel shaders, I'm not sure if @phire has any plans for touching those areas of the code base.

iwubcode commented:
Was playing around in a separate branch with DirectX12 and noticed a few errors about uninitialized values in the uber shader.  I fixed those.  Otherwise no changes.

delroth reviewed (commented):
Is the commit message meant to be "where a separate action was being created for each target"?

delroth commented on Source/Core/VideoCommon/LightingShaderGen.cpp:
```diff
@@ -175,3 +175,116 @@ void GetLightingShaderUid(LightingUidData& uid_data)
+    {
+      for (int i = 0; i < 8; ++i)
+      {
+        if ((light_mask & (1 << (i + 8 * j))) != 0)
```
Was this comment resolved at some point? It's still marked as open on GitHub.

iwubcode replied:
Whoops, I had seen this but then kept going back and forth on what to do and never commented.  I like the conciseness but also felt like the current code was a little clearer and matched some of the logic the emulator already had.  What does everyone else think?

delroth commented on Source/Core/VideoCommon/VertexManagerBase.cpp:
```diff
@@ -109,6 +109,7 @@ bool VertexManagerBase::Initialize()
   m_frame_end_event = AfterFrameEvent::Register([this] { OnEndFrame(); }, "VertexManagerBase");
   m_index_generator.Init();
   m_custom_shader_cache = std::make_unique<CustomShaderCache>();
+  m_pixel_shader_start_time = std::chrono::steady_clock::now();
```
Shouldn't that be based on the emulated clock instead?

JoshuaMKW commented:
Poke. How's the weather around here?

AdmiralCurtiss commented:
Can you break out the first two commits here? They look like they're pretty standalone and I can just merge them as-is.

AdmiralCurtiss commented:
Alright, now who do we bribe to review this one? Because this is very much outside my area of knowledge.

iwubcode commented:
Still waiting on #12040 but this is the updated version that supports all the asset work I've done in the last couple months.

There are a number of changes to the documentation.  I realize "material" and "shader" assets aren't explained that well and may be confusing.  That will likely be in some separate documentation.  This aims to set Dolphin into a place where we can write an editor (because writing this by hand, even with some python scripts I have, is still a bit painful).

Outside of the asset change, this also positions the capability to be more than a pixel shader replacement.  In the future, the hope is we can support other shader types (ex: tessellation, mesh, or geometry).  Because of this, I changed the feature from "custom pixel shader" to a "custom pipeline".

But the immediate change this provides auto-reloading of textures and shaders which makes working with this capability _much_ more enjoyable.

iwubcode commented:
All - we are aiming to merge this in early September if no one has voiced their intention to review this.  If you are interested in looking it over or have any questions, feel free to reach out (either here or Discord).  I'm happy to walk through it.

---------------------

@Pokechu22 - sorry to ping you directly but with your vast amounts of graphics emulation knowledge, parts of this review feel like a good fit.  I was going to explain everything but don't want to overload you.  The specific part of the code that I think you'd be best to look at is the part that grabs the texture information.  Most of this is currently in PixelShaderGen.

For a given draw call, the code attempts to grab all the [texture map ids](https://github.com/iwubcode/dolphin/blob/custom-shaders/Source/Core/VideoCommon/PixelShaderGen.cpp#L1789) and the [uvs](https://github.com/iwubcode/dolphin/blob/custom-shaders/Source/Core/VideoCommon/PixelShaderGen.cpp#L784).  It is attempting to give the user both the uvs and the texture sampler index for each texture.

It does this by providing an index into the 8 normal textures a game can have mapped that I call this a "texture unit". where the texture's [unit](https://github.com/iwubcode/dolphin/blob/custom-shaders/Source/Core/VideoCommon/VertexManagerBase.cpp#L590) (and ultimately [here](https://github.com/iwubcode/dolphin/blob/custom-shaders/Source/Core/VideoCommon/GraphicsModSystem/Runtime/Actions/CustomPipelineAction.cpp#L184)) will be provided to allow the user to lookup the appropriate texture sampler/uv for whatever operation the user wants to do in their custom shader.

There is definitely some duplication in getting this data, as Dolphin gets much of it for its emulation logic.  However, it's all woven in with other emulation logic (ex: the tevstage stuff) and with the review being as big as it was, I didn't want to touch it at the moment.

This review is one piece of a giant project and so I was fine merging as is for now and revisiting it later but would like to hear thoughts on structuring this (maybe a separate header for common shader operations?) or if you see any way to simplify.

Pokechu22 commented:
I will try to take a look later today - this has been on my list for a while.

Pokechu22 reviewed (commented):
(I still plan on looking over more files, but I figure it's better to bring this up and then look more into it later than have these sit around and be forgotten about)

Pokechu22 commented on Source/Core/VideoCommon/VertexManagerBase.cpp:
```diff
@@ -610,7 +640,65 @@ void VertexManagerBase::Flush()
+        case ShaderCompilationMode::SynchronousUberShaders:
+        {
+          // D3D has issues compiling large custom ubershaders
+          // use synchronous instead
```
I believe the best term here would be "specialized shaders" instead - that's the one we use on the UI, though the enum hasn't been changed for it.

Also, I'm not entirely sure why all of these use `GetPipelineAsync`. What happens if the pipeline isn't present? Does it render without the custom shaders?

iwubcode replied:
> Also, I'm not entirely sure why all of these use GetPipelineAsync. What happens if the pipeline isn't present? Does it render without the custom shaders?

Yes, if there is an error, we fall back to the game shader

Pokechu22 reviewed (commented):
OK, I did a pass over all of the shader-related code (other than `UberShaderPixel` which I didn't look at in depth - I assume it's basically the same as `PixelShaderGen`).

One thing I recommend you do is watch a few videos by Jasper, which discuss some more interesting rendering techniques used by Super Mario Galaxy and Wind Waker:

* [How scrolling textures gave Super Mario Galaxy 2 its charm](https://www.youtube.com/watch?v=f1sUPUGu9iw)
* [How The Wind Waker Redefined Cel Shading](https://www.youtube.com/watch?v=mnxs6CR6Zrk)

There are also two longer follow-up videos which discuss some other aspects of rendering:
* [A Deeper Dive into Wind Waker's Lighting - a Conversation with ScruffyMusic, part 1](https://www.youtube.com/watch?v=f1sUPUGu9iw)
* [The Charm Behind Super Mario Galaxy 2 - a Conversation with ScruffyMusic, part 2](https://www.youtube.com/watch?v=2wpQ8dVXum8)

Although these aren't really guides to doing gamecube graphics, they do give examples of more niche situations you might run into (and scenes that are worth experimenting with in games). Although on the other hand, it probably isn't as important to try and replace rendering of scenes like these if they already look good.

iwubcode commented on Source/Core/VideoCommon/ShaderGenCommon.cpp:
```diff
@@ -362,3 +363,89 @@ const char* GetInterpolationQualifier(bool msaa, bool ssaa, bool in_glsl_interfa
+  }
+  out->Write("\tuint texcoord_count;\n");
+  out->Write("\tuint[8] texmap_to_texcoord_index;\n");
+  out->Write("\tCustomShaderLightData[32] lights;\n");
```
Thanks.  Yeah, that might make more sense, though also makes it more painful for the user.  It's still not clear to me what an "alpha" light even is.  I guess it's game specific.

It's also not clear to me why there are separate "channels", again, guessing this is game specific in terms of usage.

The ideal scenario is that the user doesn't care about any of the game specific functionality and can just loop over the lights for the draw call, doing whatever custom light functionality they want.

I suppose that's still possible with that notation.  Will play around with it!

Pokechu22 replied:
Some of the Jasper videos help with this - for instance the long video on Mario Galaxy mentions that the backlighting effect is done by having a light at the camera that affects the channel 1 alpha, which is used to *decrease* the brightness of the model. It's a bit weird and doesn't physically map to a real light, but it works well in the game.

I *think* the original intent of the two channels was that one was used for diffuse lighting and the other for specular highlights, but I don't know if that's how it's generally used in practice.

iwubcode replied:
Still uncertain what the best approach here would be.  One of Jasper's video said Mario Galaxy only had 3 real "lights" which from what I've seen, didn't surprise me, but still that's so few!  I wonder if channel 0 color is the typical light channel or if it really varies based on game.

iwubcode replied:
The main thing is, it feels a little cumbersome to have to loop over both color channels and both alpha channels.  But I also don't know if it makes sense to include them.  I suppose I'm fine going with all channels split.  Really won't know what is best until we get into it.

magcius replied:
There are eight hardware lights, and in effect, four lighting channels, COLOR0, ALPHA0, COLOR1, ALPHA1. ALPHA0/ALPHA1 can be configured entirely differently from COLOR0/COLOR1 -- entirely different light masks can be used, if wanted, the only restriction is that ALPHA0/ALPHA1 are a single value instead of tricolor, and COLOR/ALPHA are bundled together in a TEV stage when selected.

If you want to, you can think of alpha lights as just ways of doing a per-vertex dot product against an arbitrary other vector. That could be used to emulate a fresnel effect if modulated in the material -- that's exactly what Super Mario Galaxy is doing, for instance.

iwubcode commented on Source/Core/VideoCommon/PixelShaderGen.cpp:
```diff
@@ -732,8 +752,135 @@ uint WrapCoord(int coord, uint wrap, int size) {{
+    }
+  }
+
+  for (u32 i = 0; i < uid_data->genMode_numindstages; ++i)
+  {
+    if ((uid_data->nIndirectStagesUsed & (1U << i)) != 0)
+    {
+      u32 texcoord = uid_data->GetTevindirefCoord(i);
+
+      // Quirk: when the tex coord is not less than the number of tex gens (i.e. the tex coord
+      // does not exist), then tex coord 0 is used (though sometimes glitchy effects happen on
+      // console). This affects the Mario portrait in Luigi's Mansion, where the developers forgot
+      // to set the number of tex gens to 2 (bug 11462).
+      if (texcoord >= uid_data->genMode_numtexgens)
+        texcoord = 0;
+
+      out->Write("\t{{\n");
+      out->Write("\t\tint2 fixpoint_uv = int2(custom_data.texcoord[{}].xy", texcoord);
+      out->Write(" * float2(" I_TEXDIMS "[{}].zw * 128));\n", texcoord);
+      out->Write("\t\tint2 tempcoord = fixpoint_uv >> " I_INDTEXSCALE "[{}].{};\n", i / 2,
+                 (i & 1) ? "zw" : "xy");
+      out->Write("\t\tcustom_data.texcoord[{0}] = float3(tempcoord / float2(" I_TEXDIMS
+                 "[{0}].zw * 128), 0);\n",
+                 texcoord);
+      out->Write("\t}}\n");
+    }
+  }
```
Yeah.  Essentially the goal is to get the texture samplers and the texture coordinates used by the game for a given draw call.  At the moment draw-call = texture, which of course isn't really right (ex: multiple textures defined for a draw) but it's hard for users to point at a draw call without some sort of editor.

The goal of this block was to grab the texture coordinates, regardless of whether it is an indirect or direct call.  This normalizes the data, so the user doesn't have to care (they probably don't).

> I'm pretty sure you can use the same texture coordinate for multiple indirect stages (e.g. both of them use texture coordinate 0 for different textures)

Yes, this is fine.  Just want the user to say for this texture, give me the sampler and texture coordinates.

> Maybe it'd be better to just get rid of this entirely for now.

I'm fine doing that but still not clear what the long-term solution would be?  If the draw-call = texture is the issue, I can certainly fix that in the future.  I vaguely recall before I had this, lots of textures just weren't mapping right.

Pokechu22 replied:
(Will address thus more fully later - I don't entirely understand what you mean by draw-call = texture, and will need to check more to understand better).

> Yes, this is fine. Just want the user to say for this texture, give me the sampler and texture coordinates.

It's fine in the case of https://github.com/dolphin-emu/dolphin/pull/11300#discussion_r1292906918, but not here where the same texture coordinate will get scaled multiple times.

Pokechu22 replied:
> The goal of this block was to grab the texture coordinates, regardless of whether it is an indirect or direct call. This normalizes the data, so the user doesn't have to care (they probably don't).

<details><summary>Explanation attempt 1</summary>

This isn't something that makes sense to do. Based on [libogc's `GX_SetIndTexCoordScale`](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L3252-L3264), the main use of the indirect scale is so that you can use the same texture coordinate for both a regular texture and an indirect texture, but have different sizes for the two.

For each texture coordinate, `bpmem.texcoords[i].s.scale_minus_1` gives a scale on the horizontal axis, and `t` on the vertical axis, which are _generally_ set to the size of the corresponding texture in texels ([libogc's `GX_SetTexCoordScaleManually`](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L4555-L4574) says that this happens automatically in most cases, but can be overridden if needed). So for a 256 by 256 texture, you would supply input coordinates between (0, 0) and (1, 1), and then these would be scaled to (0, 0) to (256, 256) when actually sampling the texture. GLSL's `texture` function doesn't expect this multiplication by the texture size, so we have to divide it by the texture's actual size ahead of time; the texture's actual size is stored in `I_TEXDIMS[i].xy` while this scale is `I_TEXDIMS[i].zw`. (Manual texture sampling uses `texelFetch` which does require multiplication by the texture size, and also manual wrapping; there's an extra complication here about the texture size the game uses being stored in `I_TEXDIMS[i].xy`, but custom textures possibly having a _different_ size that needs to be used instead, which involves querying the size in GLSL.)
</details>

... OK, that's a bit of a confusing explanation. I'm going to try it in a different way:

Games usually specify textures coordinates as UV coordinates where (0, 0) is the top-left of the texture and (1, 1) is the bottom-right of the texture. This includes the GameCube/Wii; Modern graphics APIs have functions such as `texture` where you can just provide UVs and everything works. This is probably the coordinate system you want to expose for shaders.

But for the GameCube/Wii, sampling textures is done with ST coordinates. For instance, for a 256 by 256 texture, (0, 0) is the top-left and (255, 255) is the bottom-right (... ish, there may be an off-by-one here). I've seen this UV versus ST distinction in graphics textbooks, but Dolphin doesn't really use it currently. There's a scale factor that UV coordinates get multiplied by to convert to ST coordinates. That said, this is _mostly_ hidden from the GX API: unless [`GX_SetTexCoordScaleManually`](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L4555-L4574) is used, the scale (in `bpmem.texcoords[texcoord].s.scale_minus_1` for `s`, and a similar value for `t`, and in shaders `I_TEXDIMS[texcoord].zw`) will match the one for the texture being used (in `AllTexUnits.GetUnit(texmap).texImage0` or `I_TEXDIMS[texmap].xy` - note the difference in indexing here).

This does pose an issue when using the same texture coordinate for multiple textures, if those textures are of different sizes (for instance one is 256 by 256 and the second is 128 by 128); the scale will only match one of those. If the scale factor is 256, then the 128 by 128 texture will be drawn at half-scale horizontally (as in it'll be drawn twice for every time the bigger texture is drawn, or rather 4 times since this applies on both axes). I don't think there's any mechanism to deal with this for most textures that end up getting drawn to the screen, but there _is_ a mechanism for doing this for indirect textures: [`GX_SetTexCoordScaleManually`](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L4555-L4574), which modifies `bpmem.texscale[indstage/2]`. This lets you scale down the texture coordinates when sampling a texture for an indirect texture operation (i.e. when you're reading a texture so that you can then add the values read from that texture to another texture coordinate, possibly transformed by a matrix). So, if you have a 256 by 256 main texture, and also a 128 by 128 texture you want to use as an indirect texture, you can set `bpmem.texscale[0].ss0` and `ts0` to 1, and thus the ST coordinates will be at appropriate scales for both textures.

The key issue here is that this kind of scaling is meaningless for UV coordinates, where the texture's size isn't a factor at all. I'd expect most custom shaders to use `texture` (and not e.g. `texelFetch`), so applying the scaling wouldn't make sense. Furthermore, even if ST coordinates were used, it would be more useful to provide ST coordinates on the main texture's scale, since that's what ends up on the screen eventually. (There's also a secondary issue of multiple indirect stages where all of the stages use the same texture coord; if you have two 128 by 128 indirect textures, the current code would divide s and t by 4, instead of 2.)

This is a lot of text, but hopefully it explains what the point of all of these are well enough.

iwubcode replied:
I just want to say thank you so much for taking the time to write this in-depth explanation.  I've read it a couple times and your reasoning makes a lot of sense now.  I had heard of ST coordinates but always assumed they were the same as UV coordinates.  Maybe this distinction is special to the Cube/Wii?

Regardless, I see now why I don't need the indirect texture uv transformation and only need to capture those texture map ids.

Thanks again for the explanation.  I have it all sorted now I think!

Pokechu22 replied:
I tried to dig up more information regarding ST versus UV and can't find anything concrete about the differences between them (either online or in various GC patents or in my graphics textbook), just that both are used by convention for textures. It would be convenient if there was a commonly used set of terminology for the difference between ranging between 0 and 1 versus 0 and width, but it doesn't seem like there actually is one. This doesn't change the actual GameCube/Wii behavior though (which should be as I described above); I'm just clarifying in case you see the terms elsewhere.

magcius replied:
The original difference was that UV were surface coordinates (the range is whatever you want), and ST are texture coordinates (texture matrices applied, wrap mode is applied to fit within 0-1, sometimes the coordinates are unnormalized), but this hasn't been true in practice for many many years as hardware advancements happened and as shaders took off.

UV is now the standard, ST died out as it isn't a meaningful unit unless you make GPUs.

magcius replied:
> This does pose an issue when using the same texture coordinate for multiple textures, if those textures are of different sizes (for instance one is 256 by 256 and the second is 128 by 128); the scale will only match one of those. If the scale factor is 256, then the 128 by 128 texture will be drawn at half-scale horizontally (as in it'll be drawn twice for every time the bigger texture is drawn, or rather 4 times since this applies on both axes).

The standard way to handle this is to just eat up multiple texgens from the same texture coordinate.

As mentioned, this is secretly handled for you with the GX API -- right before a display list or a draw begins, it sets the texcoord scale registers to be the size of the texture. The GX's per-pixel pipeline is *all* unnormalized, the per-vertex XF unit generates unnormalized coordinates as one of the last things it does.

This means that indirect texture offsets are unnormalized, too. If I read +128 from an indirect texture, that's going to move +128 pixels in the final unnormalized coordinate no matter the original size. But the IndTexMtx can be used to adjust for this.

> I don't think there's any mechanism to deal with this for most textures that end up getting drawn to the screen, but there is a mechanism for doing this for indirect textures: [GX_SetTexCoordScaleManually](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L4555-L4574), which modifies bpmem.texscale[indstage/2]

I think you meant to link GXSetIndTexCoordScale here?

Pokechu22 replied:
Thanks for the info on UV versus ST.

It also sounds like the distinction I wanted was normalized versus unnormalized, where normalized is for ranges of 0 to 1 while unnormalized is for ranges of 0 to width/height. (... or at least can be, I guess, and you'd also have to consider wrapping at some point, but eh.)

> I think you meant to link GXSetIndTexCoordScale here?

Yes, I meant to copy the link to [GX_SetIndTexCoordScale](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L4555-L4574) from my first abandoned explanation, but I must not have been paying attention and copied the wrong one.

> This means that indirect texture offsets are unnormalized, too. If I read +128 from an indirect texture, that's going to move +128 pixels in the final unnormalized coordinate no matter the original size. But the IndTexMtx can be used to adjust for this.

There's also an [`indtexbias`](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L802-L817) parameter to [`GX_SetTevIndirect`](https://github.com/devkitPro/libogc/blob/0a35583f638c3705be220ae5b65bd7ec20db234d/gc/ogc/gx.h#L3206-L3216) (which ends up in `bpmem.tevind[stage].bias`), which will subtract 128 from it in most cases, prior to the matrix being applied. (The exception is when the indirect format is not `ITF_8`, which is used in the Skyward Sword map (see #9876 and [my writeup](https://gist.github.com/Pokechu22/dfebe2758d23faeba48bb7c9fb8557a2).)

> The standard way to handle this is to just eat up multiple texgens from the same texture coordinate.

Yeah, that's what I've generally seen. If a game does this, then there's no special handling needed for graphics mods; an array listing the texture coordinate corresponding to each texture will work fine. (PixelShaderGen.cpp gets those generated texture coordinates as an input, so even if the model itself has only one texture coordinate or if the texture coordinates are generated from positions, at this point everything will be fine.)

> The GX's per-pixel pipeline is all unnormalized, the per-vertex XF unit generates unnormalized coordinates as one of the last things it does.

It's worth noting that Dolphin does this multiplication by texture size in the pixel shader. This is probably suboptimal since the texture size won't change, but it does mean that (apart from when per-pixel lighting is enabled) everything in the vertex shader is based on `xfmem` and everything in the pixel shader is based on `bpmem`.

[Yagcd](http://hitmen.c02.at/files/yagcd/yagcd/chap5.html#sec5.11.1) says that the relevant variables are `SU_SSIZE0` and `SU_TSIZE0`, which I think refers to the setup unit which is between the transform unit and the TEV (I think it's responsible for making triangles, although the rasterizer is also listed separately).

Handling both the multiplication by the scale in the vertex shader would be nice, but we'd still need to divide by the texture size for GLSL's `texture` and similar to work properly, both for Dolphin's use and for graphics mods. (And _that_ division couldn't be done in the vertex shader if we want fixed-point math and indirect textures to work).

iwubcode commented on Source/Core/VideoCommon/ShaderGenCommon.cpp:
```diff
@@ -362,3 +363,89 @@ const char* GetInterpolationQualifier(bool msaa, bool ssaa, bool in_glsl_interfa
+  out->Write("\tfloat3 direction;\n");
+  out->Write("\tfloat3 color;\n");
+  out->Write("\tuint attenuation_type;\n");
+  out->Write("\tuint light_type;\n");
```
Yeah it might not be.  At the moment, most of my light logic just iterates through the color lights.  I haven't done much investigation into the alpha lights.

On the (very large) todo list is a light visualizer.  I imagine once I have that, it'll be easier to understand what is going on with those four light categories.

iwubcode commented on Source/Core/VideoCommon/PixelShaderGen.cpp:
```diff
@@ -732,8 +752,135 @@ uint WrapCoord(int coord, uint wrap, int size) {{
+    out->Write("\tcustom_data.texmap_to_texcoord_index[{0}] = {0};\n", i);
+  }
+
+  // Try and do a best guess on what the texcoord index is
```
> Indirect textures don't get mapped to anything here, so they would lack an entry in texmap_to_texcoord_index. This can be fixed probably.

Yeah, if you have any ideas here, please let me know :)

> If multiple textures both use the same texture coordinate, then only the later texture coordinate is used. I don't think there's any way to fix this.

Like mentioned, if they are the same, I don't think this matters?  Maybe I am misunderstanding you.

--

Regardless, these are great points to make.  I will add them.  Thanks!

Pokechu22 replied:
Ah, right, I rethought this, and multiple textures using the same texture coordinate won't cause an issue (you'd just have e.g. `texmap_to_texcoord_index[0] = 1` and `texmap_to_texcoord_index[1] = 1` which is fine).

Instead, there is an issue if the same texture is used multiple times within a single draw. For instance, this would affect bumpmapping in Rogue Squadron ([RogueSquadron2BumpMapping](https://fifo.ci/media/dff/RogueSquadron2BumpMapping.dff)), they sample texture 4 (which is a wavy greyscale texture that acts as a height for the sand, see object 12/offset 00025aca) twice, once with texture coordinate 4 and then again with texture coordinate 5, where texture coordinate 5 is generated by adding an offset to texture coordinate 4). With the current implementation, you'll get `texmap_to_texcoord_index[4] = 5`, but I don't think there's really a better way of representing this.

Pokechu22 replied:
For indirect textures, you'd want this:

```C++
for (u32 i = 0; i < uid_data->genMode_numindstages; ++i)
{
  if ((uid_data->nIndirectStagesUsed & (1U << i)) != 0)
  {
    u32 texcoord = uid_data->GetTevindirefCoord(i);
    const u32 texmap = uid_data->GetTevindirefMap(i);

    // Quirk: when the tex coord is not less than the number of tex gens (i.e. the tex coord does
    // not exist), then tex coord 0 is used (though sometimes glitchy effects happen on console).
    // This affects the Mario portrait in Luigi's Mansion, where the developers forgot to set
    // the number of tex gens to 2 (bug 11462).
    if (texcoord >= uid_data->genMode_numtexgens)
      texcoord = 0;

    out->Write("\tcustom_data.texmap_to_texcoord_index[{}] = {};\n", texmap, texcoord);
  }
}
```

You'd probably want to do this before the current code that uses `tevstage.tevorders_texcoord` so that those values overwrite the previous values (though I don't think there is ever a situation where using a texture both as an indirect texture (what the videos referred to as a displacement map in some cases) and also as a source for colors to put on screen would make sense).

iwubcode replied:
Thank you!!  I came to the same conclusion.  I pushed that up, I did leave the tex coord modification after looking at Dolphin's logic a bit more.  Seems I can avoid the scale problem by having a separate variable.

Unfortunately, graphics mods don't (easily) work with fifo logs.  So I might punt on that until I get that investigation until I finish my editor.

iwubcode replied:
Added a comment with your updated findings.  Thank you!

iwubcode commented:
@Pokechu22 - 1'000 thank yous for reviewing this.  I'm sure it wasn't easy.

I will start looking at making the changes you suggested.

Also, I appreciate you referencing that material.  I had seen @magcius 's videos probably a handful of times, it is still hard for me to fully understand Dolphin's functionality and also map it to a more modern interface.  I probably need to watch them again.  Thank you!

iwubcode commented on Source/Core/VideoCommon/ShaderGenCommon.cpp:
```diff
@@ -362,3 +363,89 @@ const char* GetInterpolationQualifier(bool msaa, bool ssaa, bool in_glsl_interfa
+
+void WriteCustomShaderStructDef(ShaderCode* out, u32 numtexgens)
+{
+  // Roll this when there are breaking changes to the API
+  out->Write("const uint CUSTOM_SHADER_API_VERSION = 1;\n");
```
Great idea.  Thank you!

iwubcode commented on Source/Core/VideoCommon/VertexManagerBase.cpp:
```diff
@@ -523,10 +529,18 @@ void VertexManagerBase::Flush()
   // Calculate ZSlope for zfreeze
   const auto used_textures = UsedTextures();
   std::vector<std::string> texture_names;
+  std::vector<u32> texture_units;
```
Thanks.  Yeah, I will leave for now.  This will likely need to be changed in the (near) future.  I will definitely consider this then.  I guess if I use it, I'll move that to common.

iwubcode commented:
@Pokechu22 - thank you again for the review.  I've addressed the points raised.  I hope you will find some value in this feature in the future for your debugging.  At least I'd imagine this might be useful to visualizer certain scenarios.  Thanks again!

iwubcode commented:
Nothing major in this last push, just noticed a lighting shader compiler error when testing a separate feature.  Still ready for re-review.

Pokechu22 commented on docs/CustomPipelineGraphicsMod.md:
```diff
@@ -0,0 +1,251 @@
+|``position``             | vec3                    | v1    | The position of the light in _view space_                                                       |
+|``direction``            | vec3                    | v1    | The direction in _view space_ the light is pointing (only applicable for point and spot lights) |
+|``color``                | vec3                    | v1    | The color of the light                                                                          |
+|``light_type``           | uint                    | v1    | Whether the light is a color or alpha light.  See details below                                 |
```
I'm pretty sure you deleted

```markdown
|``attenuation_type``     | uint                    | v1    | The attentuation type of the light.  See details below                                          |
```

when you meant to delete this line, as `light_type` no longer exists but `attenuation_type` does.

iwubcode replied:
Dang, I did.  Thanks! :)

Pokechu22 replied:
For reference on how I noticed this:

1. `git fetch origin 5cf812458a162bd91b18f385266d1d4eba9c9f65` to get the old version of your PR
2. `git switch -d 5cf812458a162bd91b18f385266d1d4eba9c9f65` to switch to it
3. `git fetch origin 91c8cc9e4b70784a69d8967e8662a17a4f979fc0` to get the new version of your PR
4. `git log 91c8cc9e4b70784a69d8967e8662a17a4f979fc0` to find the commit from master your PR was based on (I'm sure there's an automatic way to do that, but I don't know what it is). In this case it was 1854ff01422a69bde7b2267e9ed4fc69c30ee37a.
5. `git rebase 1854ff01422a69bde7b2267e9ed4fc69c30ee37a` to rebase the old version of the PR (as a detached HEAD)
6. `git diff HEAD 91c8cc9e4b70784a69d8967e8662a17a4f979fc0` to compare the old one with the new one.

It'd be nice if github's compare changes functionality automatically handled this, but instead you also get changes in master that aren't relevant.

iwubcode replied:
I think there's some way (locally?) to do this with the ref log but I've never really looked into it.  Nice sleuthing!

Pokechu22 commented on docs/CustomPipelineGraphicsMod.md:
````diff
@@ -0,0 +1,251 @@
+
+The following shader displays the color red on the screen:
+
+```
+vec4 custom_main( in CustomShaderData data )
+{
+	return vec4(1.0, 0.0, 0.0, 1.0);
+}
+```
````
Currently this renders as:

```
vec4 custom_main( in CustomShaderData data )
{
	return vec4(1.0, 0.0, 0.0, 1.0);
}
```

If you write

<pre lang="markdown">
```glsl
</pre>

then you get syntax highlighting, like this:

```glsl
vec4 custom_main( in CustomShaderData data )
{
	return vec4(1.0, 0.0, 0.0, 1.0);
}
```

(there's also <code>```json</code> and a bunch of other languages.)

iwubcode replied:
Appreciate it, I will update those!

Pokechu22 reviewed (commented):
Apart from a minor documentation issue, this looks good to me now. I haven't done any kind of testing though.

iwubcode commented:
To @Pokechu22 and the rest of the reviewers that gave their time toward this feature, I just want to say thank you!
